pilot/pkg/networking/v1alpha3/clusters.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha3
+
+import (
+	"fmt"
+
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+
+	v2_cluster "github.com/envoyproxy/go-control-plane/envoy/api/v2/cluster"
+	"github.com/gogo/protobuf/types"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+// BuildClusters returns the list of clusters for the given proxy. This is the CDS output
+func BuildClusters(env model.Environment, node model.Proxy) []*v2.Cluster {
+	var clusters []*v2.Cluster
+
+	// TODO node type router
+
+	// node type Sidecar
+	services, err := env.Services()
+	if err != nil {
+		log.Errorf("Failed for retrieve services: %v", err)
+		return clusters
+	}
+
+
+	// build Outbound Clusters
+	for _, service := range services {
+		config := env.IstioConfigStore.DestinationRule(service.Hostname, "")
+		if config != nil {
+			destinationRule := config.Spec.(*networking.DestinationRule)
+			for _, svcPort := range service.Ports {
+
+				var hosts []*core.Address
+				if service.Resolution == model.DNSLB {
```

Comment:
did you mean model.ClientSideLB ? coz  for DNS lb, you don't need to query the registry

Reply:
@rshriram where is the DNS hostname coming from, service.Hostname?

Reply:
The code in externalservice.go seems to have this correctly. Are you writing this code from scratch, or are you reusing that file ?